docs/data/data-grid/accessibility/DensitySelectorSmallGrid.js

```diff
@@ -26,6 +26,9 @@ export default function DensitySelectorSmallGrid() {
       <DataGrid
         {...data}
         density="compact"
+        onDensityChange={(newDensity) => {
+          console.info(`Density updated to: ${newDensity}`);
```

Comment:
Should we use the controlled pattern here for consistency?
It might not be obvious that `density` is only used as an initial value and there's no way to control the density from the external state.

Reply:
Yes, that sounds better, we could even simplify the density state and move the `density.factor` computation to the selector, I did it locally (just [pushed](https://github.com/mui/mui-x/pull/12332/commits/6dd1f08e283cea60a5fa3d39ee7e26cfe2811cb0) it too), but I was not sure due to the density having a default value:
https://github.com/mui/mui-x/blob/bae7c103ecfee7b183503c77961236f841370ca4/packages/x-data-grid/src/models/props/DataGridProps.ts#L157

To follow the controlled pattern, we'd break another one 😄:
https://github.com/mui/mui-x/pull/12332/commits/6dd1f08e283cea60a5fa3d39ee7e26cfe2811cb0#diff-18991c189d0e0b75b030b0889cfbee75905b8b4c19361d0a4b975a1a3ac26945R406-R410
Note that we have to move the density prop to the `DataGridPropsWithoutDefaultValue` interface, although it technically _has_ a default value.

Plus, without making it fully controlled (i.e. wrapping it with registerControlState), the density could be updated from an external state, it's just it does not work like all other controlled models. (only works on change of prop, which could be fine since it's a primitive type and re-rending should not be an issue)
https://github.com/mui/mui-x/blob/bae7c103ecfee7b183503c77961236f841370ca4/packages/x-data-grid/src/hooks/features/density/useGridDensity.tsx#L55-L57

Wdyt?

Reply:
> To follow the controlled pattern, we'd break another one 😄:
https://github.com/mui/mui-x/commit/6dd1f08e283cea60a5fa3d39ee7e26cfe2811cb0#diff-18991c189d0e0b75b030b0889cfbee75905b8b4c19361d0a4b975a1a3ac26945R406-R410

Not sure I follow what we're breaking here.

> Note that we have to move the density prop to the DataGridPropsWithoutDefaultValue interface, although it technically has a default value.

Why?

Making the `density` a controlled state is a breaking change though (if you were passing `density="standard"`, the density selector won't work unless you handle the `onDensityChange` callback).
This also means that we need `initialState.density` to be able to initialize the density without controlling it 😅

Reply:
> Not sure I follow what we're breaking here.
Why?

Exactly why you mentioned

> This also means that we need `initialState.density`

So we can't keep initializing it here for the users to be able to pass it in `initalState.density`:
https://github.com/mui/mui-x/blob/f5dcc9d5119ba738cd2f72eb8cc1d331ceccb94a/packages/x-data-grid/src/DataGrid/useDataGridProps.ts#L29-L40

This would basically mean that we can't keep it anymore in `DataGridPropsWithDefaultValues` as all the types in it should have their default values in `const DATA_GRID_PROPS_DEFAULT_VALUES`, so we move it to `DataGridPropsWithoutDefaultValue` (unless we add an `Omit<>` to `DataGridPropsWithDefaultValues`). But we still need a fallback initial value for the Grid to keep working as expected, so we do that in the state initializer like [this](https://github.com/mui/mui-x/pull/12332/files#diff-ad2c1ad8bd2e059931470a567172fe1c91b0472508bbff7296f2572d973e08afR15).

> Making the density a controlled state is a breaking change though (if you were passing density="standard", the density selector won't work unless you handle the onDensityChange callback).

Exactly, do you think we should go in that direction though?